File: eb-application/src/main/java/com/ebnbin/ebapplication/context/ui/EBActionBarFragment.java
package com.ebnbin.ebapplication.context.ui;

import android.animation.AnimatorInflater;
import android.animation.StateListAnimator;
import android.os.Bundle;
import android.support.annotation.IdRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.AppBarLayout;
import android.support.design.widget.CollapsingToolbarLayout;
import android.support.design.widget.CoordinatorLayout;
import android.support.v4.view.NestedScrollingChild;
import android.support.v4.view.ViewCompat;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.util.ArrayMap;
import android.view.View;
import android.widget.FrameLayout;

import com.ebnbin.ebapplication.R;
import com.ebnbin.ebapplication.view.StateView;

/**
 * Base fragment with ActionBar.
 */
public abstract class EBActionBarFragment extends EBFragment {
    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        initFragmentHelper();
    }

    private CoordinatorLayout mCoordinatorLayout;
    private AppBarLayout mAppBarLayout;
    private CollapsingToolbarLayout mCollapsingToolbarLayout;
    private FrameLayout mCollapsingToolbarLayoutContentContainerFrameLayout;
    private Toolbar mToolbar;
    private FrameLayout mCoordinatorLayoutContentContainerFrameLayout;

    public CoordinatorLayout getCoordinatorLayout() {
        return mCoordinatorLayout;
    }

    public AppBarLayout getAppBarLayout() {
        return mAppBarLayout;
    }

    public CollapsingToolbarLayout getCollapsingToolbarLayout() {
        return mCollapsingToolbarLayout;
    }

    public FrameLayout getCollapsingToolbarLayoutContentContainerFrameLayout() {
        return mCollapsingToolbarLayoutContentContainerFrameLayout;
    }

    public Toolbar getToolbar() {
        return mToolbar;
    }

    public FrameLayout getCoordinatorLayoutContentContainerFrameLayout() {
        return mCoordinatorLayoutContentContainerFrameLayout;
    }

    private int mAppBarLayoutVisibleHeight;

    private StateListAnimator mDefaultAppBarStateListAnimator;
    private StateListAnimator mIgnoreExpandedAppBarStateListAnimator;

    @Override
    protected void onInitContentView(@NonNull StateView stateView, @Nullable Bundle savedInstanceState) {
        View actionBarContentContainer = getLayoutInflater().inflate(R.layout.eb_fragment_action_bar, stateView, false);
        stateView.addView(actionBarContentContainer);

        mCoordinatorLayout = actionBarContentContainer.findViewById(R.id.eb_coordinator_layout);
        mAppBarLayout = actionBarContentContainer.findViewById(R.id.eb_app_bar_layout);
        mCollapsingToolbarLayout = actionBarContentContainer.findViewById(R.id.eb_collapsing_toolbar_layout);
        mCollapsingToolbarLayoutContentContainerFrameLayout = actionBarContentContainer
                .findViewById(R.id.eb_collapsing_toolbar_layout_content_container);
        mToolbar = actionBarContentContainer.findViewById(R.id.eb_toolbar);
        mCoordinatorLayoutContentContainerFrameLayout = actionBarContentContainer
                .findViewById(R.id.eb_coordinator_layout_content_container);

        AppCompatActivity activity = getAppCompatActivity();
        if (activity != null) {
            activity.setSupportActionBar(mToolbar);
        }

        mDefaultAppBarStateListAnimator = mAppBarLayout.getStateListAnimator();
        mIgnoreExpandedAppBarStateListAnimator = AnimatorInflater.loadStateListAnimator(getContext(),
                R.animator.eb_appbar_state_list_animator_ignore_expanded);

        mAppBarLayout.addOnOffsetChangedListener(new AppBarLayout.OnOffsetChangedListener() {
            @Override
            public void onOffsetChanged(AppBarLayout appBarLayout, int verticalOffset) {
                mAppBarLayoutVisibleHeight = appBarLayout.getHeight() + verticalOffset;
                int toolbarHeight = mToolbar.getHeight();

                if (mAppBarLayoutVisibleHeight == toolbarHeight
                        && mActionBarMode == ActionBarMode.SCROLL) {
                    mAppBarLayout.setStateListAnimator(mIgnoreExpandedAppBarStateListAnimator);
                } else {
                    mAppBarLayout.setStateListAnimator(mDefaultAppBarStateListAnimator);
                }

                mActionBarModeExpanded = mAppBarLayoutVisibleHeight != 0;
            }
        });

        actionBarModeOnRestoreInstanceState(savedInstanceState);

        View contentView = overrideContentView();
        if (contentView == null) {
            int contentViewRes = overrideContentViewLayout();
            if (contentViewRes == 0) {
                return;
            }

            contentView = getLayoutInflater().inflate(contentViewRes, stateView, false);
        }

        mCoordinatorLayoutContentContainerFrameLayout.addView(contentView);
    }

    private void initFragmentHelper() {
        getFragmentHelper().setDefGroup(COORDINATOR_LAYOUT_CONTENT_CONTAINER_ID);
    }

    @Override
    protected void onChangeShared() {
        super.onChangeShared();

        AppCompatActivity activity = getAppCompatActivity();
        if (activity != null) {
            activity.setSupportActionBar(mToolbar);
        }
    }

    @IdRes
    protected static final int COLLAPSING_TOOLBAR_LAYOUT_CONTENT_CONTAINER_ID
            = R.id.eb_collapsing_toolbar_layout_content_container;
    @IdRes
    protected static final int COORDINATOR_LAYOUT_CONTENT_CONTAINER_ID = R.id.eb_coordinator_layout_content_container;

    //*****************************************************************************************************************
    // AppBarLayout scrollable.

    private boolean mAppBarCanDrag = true;

    private void setAppBarLayoutCanDrag(boolean appBarCanDrag) {
        if (mAppBarCanDrag == appBarCanDrag) {
            return;
        }

        mAppBarLayout.post(new Runnable() {
            @Override
            public void run() {
                if (!ViewCompat.isLaidOut(mAppBarLayout)) {
                    mAppBarLayout.postDelayed(this, 16L);

                    return;
                }

                CoordinatorLayout.LayoutParams params = (CoordinatorLayout.LayoutParams) mAppBarLayout
                        .getLayoutParams();
                AppBarLayout.Behavior behavior = (AppBarLayout.Behavior) params.getBehavior();
                if (behavior == null) {
                    return;
                }

                behavior.setDragCallback(new AppBarLayout.Behavior.DragCallback() {
                    @Override
                    public boolean canDrag(@NonNull AppBarLayout appBarLayout) {
                        return mAppBarCanDrag;
                    }
                });
            }
        });
    }

    private View mNestedScrollingView;
    private boolean mNestedScrollingEnabled = true;

    public void addNestedScrollingView(@NonNull View nestedScrollingView) {
        invalidateNestedScrollingEnabled(nestedScrollingView, mNestedScrollingEnabled);
    }

    public void removeNestedScrollingView(@NonNull View nestedScrollingView) {
        invalidateNestedScrollingEnabled(mNestedScrollingView == nestedScrollingView ? null : mNestedScrollingView,
                mNestedScrollingEnabled);
    }

    private void setNestedScrollingEnabled(boolean nestedScrollingEnabled) {
        invalidateNestedScrollingEnabled(mNestedScrollingView, nestedScrollingEnabled);
    }

    private void invalidateNestedScrollingEnabled(@Nullable View nestedScrollingView, boolean nestedScrollingEnabled) {
        boolean needInvalidate = false;

        if (mNestedScrollingView != nestedScrollingView) {
            mNestedScrollingView = nestedScrollingView;

            needInvalidate = true;
        }

        if (mNestedScrollingEnabled != nestedScrollingEnabled) {
            mNestedScrollingEnabled = nestedScrollingEnabled;

            needInvalidate = true;
        }

        if (!needInvalidate) {
            return;
        }

        if (nestedScrollingView instanceof NestedScrollingChild
                && !(nestedScrollingView instanceof SwipeRefreshLayout)) {
            NestedScrollingChild nestedScrollingChild = (NestedScrollingChild) nestedScrollingView;
            nestedScrollingChild.setNestedScrollingEnabled(mNestedScrollingEnabled);
        }
    }

    //*****************************************************************************************************************
    // AppBarLayout animation duration.

    private static final long APP_BAR_LAYOUT_ANIMATION_DURATION_STANDARD = 300L;
    private static final long APP_BAR_LAYOUT_ANIMATION_DURATION_MAX = 600L;

    //*****************************************************************************************************************
    // ActionBar mode.

    public enum ActionBarMode {
        STANDARD,
        SCROLL
    }

    private ActionBarMode mActionBarMode = ActionBarMode.STANDARD;

    private boolean mActionBarModeExpanded = true;

    public ActionBarMode getActionBarMode() {
        return mActionBarMode;
    }

    public void setActionBarMode(ActionBarMode actionBarMode, boolean forceInvalidate, @Nullable Boolean expanded,
            boolean animate) {
        if (mSetCollapsingToolbarLayoutScrollFlagsRunnable != null) {
            mSetCollapsingToolbarLayoutScrollFlagsRunnable.run();
        }

        if (!forceInvalidate && mActionBarMode == actionBarMode) {
            return;
        }

        if (expanded != null) {
            mActionBarModeExpanded = expanded;
        }

        mAppBarLayout.setExpanded(mActionBarModeExpanded, animate);

        mSetCollapsingToolbarLayoutScrollFlagsRunnable = new SetCollapsingToolbarLayoutScrollFlagsRunnable(
                actionBarMode);


        ActionBarModeConstants actionBarModeConstants = ACTION_BAR_MODE_CONSTANTS_ARRAY_MAP.get(actionBarMode);

        mCollapsingToolbarLayoutContentContainerFrameLayout.setVisibility(
                actionBarModeConstants.collapsingToolbarLayoutContentContainerFrameLayoutVisibility);
        setAppBarLayoutCanDrag(actionBarModeConstants.appBarLayoutCanDrag);
        setNestedScrollingEnabled(actionBarModeConstants.nestedScrollingEnabled);

        if (animate) {
            mCollapsingToolbarLayout.postDelayed(mSetCollapsingToolbarLayoutScrollFlagsRunnable,
                    APP_BAR_LAYOUT_ANIMATION_DURATION_STANDARD);
        } else {
            mSetCollapsingToolbarLayoutScrollFlagsRunnable.run();
        }
    }

    private void invalidateToolbar() {
        AppCompatActivity activity = getAppCompatActivity();
        if (activity != null) {
            activity.setSupportActionBar(mToolbar);
        }
    }

    //*****************************************************************************************************************
    // SetCollapsingToolbarLayoutScrollFlagsRunnable.

    private SetCollapsingToolbarLayoutScrollFlagsRunnable mSetCollapsingToolbarLayoutScrollFlagsRunnable;

    private class SetCollapsingToolbarLayoutScrollFlagsRunnable implements Runnable {
        private final ActionBarMode mActionBarMode;

        public SetCollapsingToolbarLayoutScrollFlagsRunnable(ActionBarMode actionBarMode) {
            mActionBarMode = actionBarMode;
        }

        @Override
        public void run() {
            mAppBarLayout.removeCallbacks(this);

            mSetCollapsingToolbarLayoutScrollFlagsRunnable = null;

            ActionBarModeConstants actionBarModeConstants = ACTION_BAR_MODE_CONSTANTS_ARRAY_MAP.get(mActionBarMode);

            AppBarLayout.LayoutParams params = (AppBarLayout.LayoutParams) mCollapsingToolbarLayout.getLayoutParams();
            params.setScrollFlags(actionBarModeConstants.collapsingToolbarLayoutScrollFlags);

            EBActionBarFragment.this.mActionBarMode = mActionBarMode;

            invalidateToolbar();
        }
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();

        if (mSetCollapsingToolbarLayoutScrollFlagsRunnable != null) {
            mSetCollapsingToolbarLayoutScrollFlagsRunnable.run();
        }
    }

    //*****************************************************************************************************************
    // Instance state.

    private static final String INSTANCE_STATE_ACTION_BAR_MODE = "action_bar_mode";
    private static final String INSTANCE_STATE_ACTION_BAR_MODE_EXPANDED = "action_bar_mode_expanded";

    private void actionBarModeOnRestoreInstanceState(@Nullable Bundle savedInstanceState) {
        if (savedInstanceState == null) {
            setActionBarMode(ActionBarMode.STANDARD, true, true, false);
            return;
        }

        ActionBarMode actionBarMode = (ActionBarMode) savedInstanceState.getSerializable(INSTANCE_STATE_ACTION_BAR_MODE);
        boolean actionBarModeExpanded = savedInstanceState.getBoolean(INSTANCE_STATE_ACTION_BAR_MODE_EXPANDED);

        setActionBarMode(actionBarMode, true, actionBarModeExpanded, false);
    }

    private void actionBarModeOnSaveInstanceState(@Nullable Bundle outState) {
        if (outState == null) {
            return;
        }

        outState.putSerializable(INSTANCE_STATE_ACTION_BAR_MODE, mActionBarMode);
        outState.putBoolean(INSTANCE_STATE_ACTION_BAR_MODE_EXPANDED, mActionBarModeExpanded);
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);

        actionBarModeOnSaveInstanceState(outState);
    }

    //*****************************************************************************************************************
    // ActionBarMode constants.

    private static final ArrayMap<ActionBarMode, ActionBarModeConstants> ACTION_BAR_MODE_CONSTANTS_ARRAY_MAP
            = new ArrayMap<>();

    static {
        int standardCollapsingToolbarLayoutScrollFlags = AppBarLayout.LayoutParams.SCROLL_FLAG_SCROLL
                | AppBarLayout.LayoutParams.SCROLL_FLAG_EXIT_UNTIL_COLLAPSED;
        ActionBarModeConstants standardActionBarModeConstants = new ActionBarModeConstants(
                standardCollapsingToolbarLayoutScrollFlags, false, false, View.GONE);
        ACTION_BAR_MODE_CONSTANTS_ARRAY_MAP.put(ActionBarMode.STANDARD, standardActionBarModeConstants);

        int scrollCollapsingToolbarLayoutScrollFlags = AppBarLayout.LayoutParams.SCROLL_FLAG_SCROLL
                | AppBarLayout.LayoutParams.SCROLL_FLAG_ENTER_ALWAYS
                | AppBarLayout.LayoutParams.SCROLL_FLAG_SNAP;
        ActionBarModeConstants scrollActionBarModeConstants = new ActionBarModeConstants(
                scrollCollapsingToolbarLayoutScrollFlags, false, true, View.GONE);
        ACTION_BAR_MODE_CONSTANTS_ARRAY_MAP.put(ActionBarMode.SCROLL, scrollActionBarModeConstants);
    }

    private static final class ActionBarModeConstants {
        public final int collapsingToolbarLayoutScrollFlags;
        public final boolean appBarLayoutCanDrag;
        public final boolean nestedScrollingEnabled;
        public final int collapsingToolbarLayoutContentContainerFrameLayoutVisibility;

        private ActionBarModeConstants(int collapsingToolbarLayoutScrollFlags, boolean appBarLayoutCanDrag,
                boolean nestedScrollingEnabled, int collapsingToolbarLayoutContentContainerFrameLayoutVisibility) {
            this.collapsingToolbarLayoutScrollFlags = collapsingToolbarLayoutScrollFlags;
            this.appBarLayoutCanDrag = appBarLayoutCanDrag;
            this.nestedScrollingEnabled = nestedScrollingEnabled;
            this.collapsingToolbarLayoutContentContainerFrameLayoutVisibility
                    = collapsingToolbarLayoutContentContainerFrameLayoutVisibility;
        }
    }
}

File: eb-application/src/main/java/com/ebnbin/ebapplication/context/ui/EBActionBarFragment.kt
package com.ebnbin.ebapplication.context.ui

import android.animation.AnimatorInflater
import android.os.Bundle
import android.support.annotation.IdRes
import android.support.design.widget.AppBarLayout
import android.support.design.widget.CollapsingToolbarLayout
import android.support.design.widget.CoordinatorLayout
import android.support.v4.util.ArrayMap
import android.support.v4.view.NestedScrollingChild
import android.support.v4.view.ViewCompat
import android.support.v4.widget.SwipeRefreshLayout
import android.support.v7.widget.Toolbar
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import com.ebnbin.eb.util.EBUtil
import com.ebnbin.ebapplication.R
import com.ebnbin.ebapplication.view.StateView

/**
 * Base [EBFragment] with ActionBar.
 */
abstract class EBActionBarFragment : EBFragment() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        initFragmentHelper()
    }

    private fun initFragmentHelper() {
        fragmentHelper.defGroup = coordinatorLayoutContentId
    }

    //*****************************************************************************************************************
    // Content view.

    private var actionBarContainer: ViewGroup? = null

    val coordinatorLayout: CoordinatorLayout by lazy {
        actionBarContainer!!.findViewById<CoordinatorLayout>(R.id.eb_coordinator_layout)
    }

    val appBarLayout: AppBarLayout by lazy {
        actionBarContainer!!.findViewById<AppBarLayout>(R.id.eb_app_bar_layout)
    }

    val collapsingToolbarLayout: CollapsingToolbarLayout by lazy {
        actionBarContainer!!.findViewById<CollapsingToolbarLayout>(R.id.eb_collapsing_toolbar_layout)
    }

    val collapsingToolbarLayoutContentFrameLayout: FrameLayout by lazy {
        actionBarContainer!!.findViewById<FrameLayout>(R.id.eb_collapsing_toolbar_layout_content)
    }

    val toolbar: Toolbar by lazy {
        actionBarContainer!!.findViewById<Toolbar>(R.id.eb_toolbar)
    }

    val coordinatorLayoutContentFrameLayout: FrameLayout by lazy {
        actionBarContainer!!.findViewById<FrameLayout>(R.id.eb_coordinator_layout_content)
    }

    private val defAppBarLayoutStateListAnimator by lazy {
        appBarLayout.stateListAnimator
    }

    private val ignoreExpandedAppBarLayoutStateListAnimator by lazy {
        AnimatorInflater.loadStateListAnimator(context,
                R.animator.eb_app_bar_layout_state_list_animator_ignore_expanded)
    }

    private var appBarLayoutExpanded: Boolean = true

    val onOffsetChangedListener: AppBarLayout.OnOffsetChangedListener by lazy {
        // Init first.
        defAppBarLayoutStateListAnimator

        AppBarLayout.OnOffsetChangedListener { appBarLayout, verticalOffset ->
            val visibleHeight = appBarLayout.height + verticalOffset

            if (visibleHeight == toolbar.height && actionBarMode == ActionBarMode.SCROLL) {
                appBarLayout.stateListAnimator = ignoreExpandedAppBarLayoutStateListAnimator
            } else {
                appBarLayout.stateListAnimator = defAppBarLayoutStateListAnimator
            }

            appBarLayoutExpanded = visibleHeight != 0
        }
    }

    override fun onInitContentView(stateView: StateView, savedInstanceState: Bundle?) {
        actionBarContainer = layoutInflater.inflate(R.layout.eb_fragment_action_bar, stateView, true)!! as ViewGroup

        appBarLayout.addOnOffsetChangedListener(onOffsetChangedListener)

        initActionBarMode(savedInstanceState)

        var contentView = overrideContentView()
        if (contentView == null) {
            val contentViewLayout = overrideContentViewLayout()
            if (contentViewLayout == 0) return

            contentView = layoutInflater.inflate(contentViewLayout, coordinatorLayoutContentFrameLayout, false)
        }

        coordinatorLayoutContentFrameLayout.addView(contentView)
    }

    override fun onDestroyView() {
        if (postActionBarModeRunnable != null) {
            postActionBarModeRunnable!!.run()
        }

        appBarLayout.removeOnOffsetChangedListener(onOffsetChangedListener)

        super.onDestroyView()
    }

    //*****************************************************************************************************************
    // AppBarLayout can drag.

    private var appBarLayoutCanDrag: Boolean = true

    private fun setAppBarLayoutCanDrag(appBarLayoutCanDrag: Boolean) {
//        if (this.appBarLayoutCanDrag == appBarLayoutCanDrag) return

        this.appBarLayoutCanDrag = appBarLayoutCanDrag

        EBUtil.handler.post { object : Runnable {
            override fun run() {
                if (!ViewCompat.isLaidOut(appBarLayout)) {
                    EBUtil.handler.postDelayed(this, 16L)

                    return
                }

                val behavior = (appBarLayout.layoutParams as CoordinatorLayout.LayoutParams)
                        .behavior as AppBarLayout.Behavior? ?: return
                behavior.setDragCallback(object : AppBarLayout.Behavior.DragCallback() {
                    override fun canDrag(appBarLayout: AppBarLayout): Boolean {
                        return this@EBActionBarFragment.appBarLayoutCanDrag
                    }
                })
            }
        } }
    }

    //*****************************************************************************************************************
    // Nested scrolling.

    private var nestedScrollingChild: NestedScrollingChild? = null
    private var nestedScrollingEnabled = false

    fun setNestedScrollingChild(nestedScrollingChild: NestedScrollingChild?) {
        invalidateNestedScrolling(nestedScrollingChild, nestedScrollingEnabled)
    }

    private fun setNestedScrollingEnabled(nestedScrollingEnabled: Boolean) {
        invalidateNestedScrolling(nestedScrollingChild, nestedScrollingEnabled)
    }

    private fun invalidateNestedScrolling(nestedScrollingChild: NestedScrollingChild?,
            nestedScrollingEnabled: Boolean) {
        var needInvalidate = false

//        if (this.nestedScrollingChild !== nestedScrollingChild) {
            this.nestedScrollingChild = nestedScrollingChild
//
//            needInvalidate = true
//        }

//        if (this.nestedScrollingEnabled != nestedScrollingEnabled) {
            this.nestedScrollingEnabled = nestedScrollingEnabled

            needInvalidate = true
//        }

        if (needInvalidate
                && this.nestedScrollingChild != null
                && this.nestedScrollingChild !is SwipeRefreshLayout) {
            this.nestedScrollingChild!!.isNestedScrollingEnabled = this.nestedScrollingEnabled
        }
    }

    //*****************************************************************************************************************
    // ActionBar mode.

    enum class ActionBarMode {
        STANDARD,
        SCROLL
    }

    private var actionBarMode = ActionBarMode.STANDARD

    fun setActionBarMode(actionBarMode: ActionBarMode, forceInvalidate: Boolean, expanded: Boolean?,
            animate: Boolean) {
        if (!forceInvalidate && this.actionBarMode == actionBarMode) return

        postActionBarModeRunnable?.run()

        if (expanded != null) {
            appBarLayoutExpanded = expanded
        }

        appBarLayout.setExpanded(appBarLayoutExpanded, animate)

        val actionBarModeConstant = actionBarModeConstants[actionBarMode]!!

        setAppBarLayoutCanDrag(actionBarModeConstant.appBarLayoutCanDrag)

        postActionBarModeRunnable = PostActionBarModeRunnable(actionBarMode)
        if (animate) {
            EBUtil.handler.postDelayed(postActionBarModeRunnable!!,
                    actionBarModeConstant.appBarLayoutAnimationDuration)
        } else{
            postActionBarModeRunnable!!.run()
        }

        setNestedScrollingEnabled(actionBarModeConstant.nestedScrollingEnabled)
    }

    private fun initActionBarMode(savedInstanceState: Bundle?) {
        if (savedInstanceState == null) {
            setActionBarMode(ActionBarMode.STANDARD, true, true, false)

            return
        }

        val actionBarMode = savedInstanceState.getSerializable(INSTANCE_STATE_ACTION_BAR_MODE) as ActionBarMode
        val appBarLayoutExpanded = savedInstanceState.getBoolean(INSTANCE_STATE_APP_BAR_LAYOUT_EXPANDED)

        setActionBarMode(actionBarMode, true, appBarLayoutExpanded, false)
    }

    private fun actionBarModeOnSaveInstanceState(outState: Bundle?) {
        if (outState == null) return

        outState.putSerializable(INSTANCE_STATE_ACTION_BAR_MODE, actionBarMode)
        outState.putBoolean(INSTANCE_STATE_APP_BAR_LAYOUT_EXPANDED, appBarLayoutExpanded)

        // TODO appBarLayoutCanDrag, nestedScrollingChild, nestedScrollingEnabled
    }

    //*****************************************************************************************************************

    private var postActionBarModeRunnable: PostActionBarModeRunnable? = null

    private inner class PostActionBarModeRunnable(private val actionBarMode: ActionBarMode) : Runnable {
        override fun run() {
            EBUtil.handler.removeCallbacks(this)

            postActionBarModeRunnable = null

            val actionBarModeConstant = actionBarModeConstants[actionBarMode]!!

            val params = collapsingToolbarLayout.layoutParams as AppBarLayout.LayoutParams
            params.scrollFlags = actionBarModeConstant.collapsingToolbarLayoutScrollFlags

            collapsingToolbarLayoutContentFrameLayout.visibility =
                    if (actionBarModeConstant.collapsingToolbarLayoutContentFrameLayoutVisible) View.VISIBLE
                    else View.GONE

            this@EBActionBarFragment.actionBarMode = actionBarMode

            appCompatActivity.setSupportActionBar(toolbar)
        }
    }

    //*****************************************************************************************************************

    override fun onSaveInstanceState(outState: Bundle?) {
        super.onSaveInstanceState(outState)

        actionBarModeOnSaveInstanceState(outState)
    }

    //*****************************************************************************************************************

    override fun onFront() {
        super.onFront()

        appCompatActivity.setSupportActionBar(toolbar)
    }

    //*****************************************************************************************************************

    companion object {
        @IdRes protected val collapsingToolbarLayoutContentId = R.id.eb_collapsing_toolbar_layout_content
        @IdRes protected val coordinatorLayoutContentId = R.id.eb_coordinator_layout_content

        //*************************************************************************************************************
        // AppBarLayout animation duration.

        private val APP_BAR_LAYOUT_ANIMATION_DURATION_STANDARD = 300L
        private val APP_BAR_LAYOUT_ANIMATION_DURATION_MAX = 600L

        //*************************************************************************************************************
        // ActionBarMode constants.

        private val actionBarModeConstants = ArrayMap<ActionBarMode, ActionBarModeConstant>()

        private data class ActionBarModeConstant(val appBarLayoutCanDrag: Boolean,
                val appBarLayoutAnimationDuration: Long, val collapsingToolbarLayoutScrollFlags: Int,
                val collapsingToolbarLayoutContentFrameLayoutVisible: Boolean, val nestedScrollingEnabled: Boolean)

        init {
            actionBarModeConstants.put(ActionBarMode.STANDARD, ActionBarModeConstant(false, 0L,
                    AppBarLayout.LayoutParams.SCROLL_FLAG_SCROLL
                            or AppBarLayout.LayoutParams.SCROLL_FLAG_EXIT_UNTIL_COLLAPSED, false, false))
            actionBarModeConstants.put(ActionBarMode.SCROLL, ActionBarModeConstant(false,
                    APP_BAR_LAYOUT_ANIMATION_DURATION_STANDARD, AppBarLayout.LayoutParams.SCROLL_FLAG_SCROLL
                            or AppBarLayout.LayoutParams.SCROLL_FLAG_ENTER_ALWAYS
                            or AppBarLayout.LayoutParams.SCROLL_FLAG_SNAP, false, true))
        }

        //*****************************************************************************************************************
        // Instance state.

        private val INSTANCE_STATE_ACTION_BAR_MODE = "action_bar_mode"
        private val INSTANCE_STATE_APP_BAR_LAYOUT_EXPANDED = "app_bar_layout_expanded"
    }
}
